src/common/github-flavored-markdown.js

```diff
@@ -1,42 +1,82 @@
-import memoizee from 'memoizee';
 import MarkdownIt from 'markdown-it';
-import replaceLink from 'markdown-it-replace-link';
+import { escapeHtml } from '../utils/string-utils.js';
 
-const getGfmForCdnRootUrl = memoizee(cdnRootUrl => {
-  let gfm;
-  if (cdnRootUrl) {
-    gfm = new MarkdownIt({
-      replaceLink: link => link.startsWith('cdn://') ? `${cdnRootUrl}/${link.substr(6)}` : link
-    });
-    gfm.use(replaceLink);
-  } else {
-    gfm = new MarkdownIt();
+const audioUrlPattern = new RegExp(`\\.(${['aac', 'm4a', 'mp3', 'oga', 'ogg', 'wav'].join('|')})$`, 'i');
+const videoUrlPattern = new RegExp(`\\.(${['mp4', 'm4v', 'ogv', 'webm', 'mpg', 'mpeg'].join('|')})$`, 'i');
+
+function getMediaType(url) {
+  if (audioUrlPattern.test(url)) {
+    return 'audio';
   }
+  if (videoUrlPattern.test(url)) {
+    return 'video';
+  }
+  return 'image';
+}
+
+function educanduFeatures(md) {
+  md.core.ruler.after('inline', 'educandu-features', state => {
+    const { cdnRootUrl, renderMedia, collectCdnUrl } = state.env;
+
+    state.tokens.filter(parentToken => parentToken.type === 'inline').forEach(parentToken => {
+      parentToken.children?.forEach(token => {
+        if (token.type === 'link_open' || token.type === 'image') {
+          let targetUrl = null;
+          const attrToReplace = token.type === 'link_open' ? 'href' : 'src';
+          token.attrs.forEach(attr => {
```

Comment:
Could `token.attrs.forEach` be replaced with a `token.attrs.find(attr => attr[0] === attrToReplace);`?
Or maybe I got this code wrong, super hard to follow 😳

Reply:
True, it saves one level of nesting, done, thanks!

Reply:
I could boil it down using the functions provided by the node class in markdown-it, so I think it looks much more understandable now. Thanks for the input!

Reply:
Looks great, thank you, now the code speaks for itself! 🙏